src/stan/model/indexing/rvalue.hpp

```diff
@@ -234,16 +239,18 @@ inline T rvalue(
  * @param[in] depth Depth of indexing dimension.
  * @return Result of indexing matrix.
  */
-template <typename T, typename I>
-inline typename boost::disable_if<boost::is_same<I, index_uni>,
-                                  Eigen::Matrix<T, 1, Eigen::Dynamic> >::type
-rvalue(
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& a,
-    const cons_index_list<index_uni, cons_index_list<I, nil_index_list> >& idx,
+template <typename EigMat, typename I,
+          typename = require_not_same_t<index_uni, I>,
+          typename = require_eigen_t<EigMat>,
+          typename = require_not_eigen_vector_t<EigMat>>
+inline auto rvalue(
+    const EigMat& a,
+    const cons_index_list<index_uni, cons_index_list<I, nil_index_list>>& idx,
     const char* name = "ANON", int depth = 0) {
   int m = idx.head_.n_;
   math::check_range("matrix[uni,multi] indexing, row", name, a.rows(), m);
-  Eigen::Matrix<T, 1, Eigen::Dynamic> r = a.row(m - 1);
+  Eigen::Matrix<scalar_type_t<EigMat>, 1, EigMat::ColsAtCompileTime> r
```

Comment:
We can avoid one eval here.
```suggestion
  const auto& r
```

Reply:
What about `decltype(auto)` here?

Reply:
I think that should work as well.